File: app/src/main/java/com/yfbx/widgets/widgets/LoadingView.java
package com.yfbx.widgets.widgets;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.SweepGradient;
import android.util.AttributeSet;
import android.view.View;

/**
 * Author: Edward
 * Date: 2018/7/16
 * Description:
 */


public class LoadingView extends View {

    private static final String TAG = "CircleAnim";

    private Paint paint;
    private int degree;
    private int width;
    private int height;
    private boolean loading;

    public LoadingView(Context context) {
        this(context, null);
    }

    public LoadingView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public LoadingView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    /**
     * 初始化
     */
    private void init() {

        int[] colors = {0xFF231EFB, 0xFF8710F8, 0xFFCB05FF};
        SweepGradient sweepGradient = new SweepGradient(500, 400, colors, null);

        paint = new Paint();
        paint.setStrokeCap(Paint.Cap.ROUND);
        paint.setStyle(Paint.Style.STROKE);
        paint.setStrokeWidth(15);
        paint.setShader(sweepGradient);
        paint.setAntiAlias(true);

    }


    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);

        width = 128;
        height = 128;
        setMeasuredDimension(width, height);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        canvas.drawArc(8, 8, 120, 120, degree, 300, false, paint);
        if (loading) {
            setRotate();
        }
    }

    private void setRotate() {
        degree += 10;
        if (degree >= 360) {
            degree = 0;
        }
        invalidate();
    }

    public void start() {
        loading = true;
        invalidate();
    }

    public void stop() {
        loading = false;
    }

    public boolean isLoading() {
        return loading;
    }

}

File: app/src/main/java/com/yfbx/widgets/widget/anim/LoadingView.kt
package com.yfbx.widgets.widget.anim

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.SweepGradient
import android.util.AttributeSet
import android.view.View

/**
 * Author: Edward
 * Date: 2018/7/16
 * Description: 绘制动画
 */
class LoadingView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : View(context, attrs, defStyleAttr) {

    private var paint = Paint()
    private var degree: Int = 0

    init {
        val colors = intArrayOf(-0xdce105, -0x78ef08, -0x34fa01)
        val sweepGradient = SweepGradient(500f, 400f, colors, null)

        paint.strokeCap = Paint.Cap.ROUND
        paint.style = Paint.Style.STROKE
        paint.strokeWidth = 15f
        paint.shader = sweepGradient
        paint.isAntiAlias = true
    }


    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        setMeasuredDimension(128, 128)
    }

    override fun onDraw(canvas: Canvas) {
        canvas.drawArc(8f, 8f, 120f, 120f, degree.toFloat(), 300f, false, paint)
        setRotate()
    }

    private fun setRotate() {
        degree += 10
        if (degree >= 360) {
            degree = 0
        }
        invalidate()
    }
}
